packages/adblocker/src/engine/bucket/html.ts

```diff
@@ -0,0 +1,213 @@
+import NetworkFilter from '../../filters/network.js';
+import CosmeticFilter from '../../filters/cosmetic.js';
+import Request from '../../request.js';
+import { noopOptimizeNetwork, optimizeNetwork, noopOptimizeCosmetic } from '../optimizer.js';
+import ReverseIndex from '../reverse-index.js';
+import { createLookupTokens } from './cosmetic.js';
+
+export default class HTMLBucket {
+  public static deserialize(buffer: StaticDataView, config: Config): HTMLBucket {
+    const bucket = new HTMLBucket({ config });
+
+    bucket.networkIndex = ReverseIndex.deserialize(
+      buffer,
+      NetworkFilter.deserialize,
+      config.enableOptimizations ? optimizeNetwork : noopOptimizeNetwork,
+      config,
+    );
+
+    bucket.exceptionsIndex = ReverseIndex.deserialize(
+      buffer,
+      NetworkFilter.deserialize,
+      config.enableOptimizations ? optimizeNetwork : noopOptimizeNetwork,
+      config,
+    );
+
+    bucket.cosmeticIndex = ReverseIndex.deserialize(
+      buffer,
+      CosmeticFilter.deserialize,
+      noopOptimizeCosmetic,
+      config,
+    );
+
+    bucket.unhideIndex = ReverseIndex.deserialize(
+      buffer,
+      CosmeticFilter.deserialize,
+      noopOptimizeCosmetic,
+      config,
+    );
+
+    return bucket;
+  }
+
+  public networkIndex: ReverseIndex<NetworkFilter>;
+  public exceptionsIndex: ReverseIndex<NetworkFilter>;
+  public cosmeticIndex: ReverseIndex<CosmeticFilter>;
+  public unhideIndex: ReverseIndex<CosmeticFilter>;
+
+  constructor({
+    filters = [],
+    config,
+  }: {
+    filters?: (CosmeticFilter | NetworkFilter)[];
+    config: Config;
+  }) {
+    this.networkIndex = new ReverseIndex({
+      config,
+      deserialize: NetworkFilter.deserialize,
+      filters: [],
+      optimize: config.enableOptimizations ? optimizeNetwork : noopOptimizeNetwork,
+    });
+
+    this.exceptionsIndex = new ReverseIndex({
+      config,
+      deserialize: NetworkFilter.deserialize,
+      filters: [],
+      optimize: config.enableOptimizations ? optimizeNetwork : noopOptimizeNetwork,
+    });
+
+    this.cosmeticIndex = new ReverseIndex({
+      config,
+      deserialize: CosmeticFilter.deserialize,
+      filters: [],
+      optimize: noopOptimizeCosmetic,
+    });
+
+    this.unhideIndex = new ReverseIndex({
+      config,
+      deserialize: CosmeticFilter.deserialize,
+      filters: [],
+      optimize: noopOptimizeCosmetic,
+    });
+
+    if (filters.length !== 0) {
+      this.update(filters, undefined);
+    }
+  }
+
+  public update(
+    newFilters: (NetworkFilter | CosmeticFilter)[],
+    removedFilters: Set<number> | undefined,
+  ): void {
+    const networkFilters: NetworkFilter[] = [];
+    const exceptionFilters: NetworkFilter[] = [];
+    const cosmeticFilters: CosmeticFilter[] = [];
+    const unhideFilters: CosmeticFilter[] = [];
+
+    for (const filter of newFilters) {
+      if (filter.isNetworkFilter()) {
+        const networkFilter = filter as NetworkFilter;
```

Comment:
Is this case needed? Maybe we can add a type guard in the `isNetworkFilter` function?

Reply:
Fixed in [`f961220` (#3897)](https://github.com/ghostery/adblocker/pull/3897/commits/f96122054d679cb48cb8fe9165592547e7a4b158)